src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -79,134 +69,206 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
+            string? bestSDK = GetSdkFromGlobalSettings(workingDirectory);
+            if (!string.IsNullOrEmpty(bestSDK))
+                yield return bestSDK;
+
+            string[] dotnetPaths = GetAllAvailableSDKs();
+            // We want to return the newest SDKs first, however, so iterate over the list in reverse order.
+            // If basePath is disqualified because it was later
+            // than the runtime version, this ensures that RegisterDefaults will return the latest valid
+            // SDK instead of the earliest installed.
+            for (int i = dotnetPaths.Length - 1; i >= 0; i--)
+            {
+                if (dotnetPaths[i] != bestSDK)
+                    yield return dotnetPaths[i];
+            }
         }
 
-        private static string FindDotnetFromEnvironmentVariable(string environmentVariable, string exeName)
+        private static void LoadHostFxr()
         {
-            string dotnet_root = Environment.GetEnvironmentVariable(environmentVariable);
-            if (!string.IsNullOrEmpty(dotnet_root))
+            var isOSX = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
+            if (isOSX)
             {
-                string fullPathToDotnetFromRoot = Path.Combine(dotnet_root, exeName);
-                if (File.Exists(fullPathToDotnetFromRoot))
-                {
-                    if (!IsWindows)
-                    {
-                        fullPathToDotnetFromRoot = realpath(fullPathToDotnetFromRoot) ?? fullPathToDotnetFromRoot;
-                        return File.Exists(fullPathToDotnetFromRoot) ? Path.GetDirectoryName(fullPathToDotnetFromRoot) : null;
-                    }
-
-                    return dotnet_root;
-                }
+                var loadContext = System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+                if (loadContext != null)
+                    loadContext.ResolvingUnmanagedDll += HostFxrResolver;
             }
-
-            return null;
         }
 
-        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
+        private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
         {
-            string dotnetPath = null;
-            string exeName = IsWindows ? "dotnet.exe" : "dotnet";
+            var hostFxrLibName = "libhostfxr.dylib";
 
-            // First check for the DOTNET_ROOT environment variable, as it's often there as with, for example, dotnet format.
-            if (IntPtr.Size == 4)
-            {
-                // 32-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT(x86)", exeName);
-            }
-            else if (IntPtr.Size == 8)
-            {
-                // 64-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT", exeName);
-            }
+            if (!hostFxrLibName.Contains(libraryName, StringComparison.Ordinal))
```

Comment:
Why contains instead of equals? Should we be stricter and check that `libraryName` is `libhostfxr`, case insensitive?

Reply:
done